virttest/utils_zcrypt.py

```diff
@@ -0,0 +1,378 @@
+        e, o = cmd_status_output("lszcrypt -V", shell=True,
+                                 session=self.session,
+                                 timeout=CMD_TIMEOUT,
+                                 verbose=VERBOSE)
+        if e:
+            if NO_DEVICES not in o:
+                raise OSError("Error when running lszcrypt: %s" % o)
+        else:
+            o = o.strip().split('\n')[2:]
+            for entry in o:
+                info.append(entry)
+        return info
+
+    @staticmethod
+    def get(session=None):
+        """
+        Gets crypto device info
+
+        :param session: guest session, s. __init__ for details
+        :return: CryptoDeviceInfo instance
+        """
+        builder = CryptoDeviceInfoBuilder(session)
+        return builder.get_info()
+
+
+class CryptoDeviceInfo(object):
+    """
+    Represents the crypto device info
+    """
+
+    def __init__(self):
+        self._entries = []
+
+    def append(self, lszcrypt_entry):
+        """
+        Append a new row on the table
+
+        :param lszcrypt_entry: LszcryptRow instance
+        :return: None
+        """
+        self._entries.append(CryptoDeviceInfoEntry.from_string(lszcrypt_entry))
+
+    @property
+    def entries(self):
+        """
+        Protect member from write
+
+        :return: crypto device info entries
+        """
+        return self._entries
+
+    @property
+    def domains(self):
+        """
+        Returns only domain info entries
+
+        :return: list of domain info entries
+        """
+        return [x for x in self._entries if x.domain]
+
+    def __str__(self):
+        s = "{_entries: "
+        for e in self._entries:
+            s += str(e.__dict__)
+        s += "}"
+        return s
+
+
+class CryptoDeviceInfoEntry(object):
+    """
+    Represents a single device info entry
+    """
+
+    def __init__(self):
+        self._card = None
+        self._domain = None
+        self.type = None
+        self.mode = None
+        self.status = None
+        self.requests = None
+        self.pending = None
+        self.hwtype = None
+        self.qdepth = None
+        self.functions = None
+        self.driver = None
+
+    @property
+    def id(self):
+        """
+        The id value as list [card], or [card, domain]
+        :return: id value as list
+        """
+        return [self.card, self.domain] if self.domain else [self.card]
+
+    @id.setter
+    def id(self, id):
+        if "." in id:
+            parts = id.split(".")
+            self._card, self._domain = parts[0], parts[1]
+        else:
+            self._card, self._domain = id, None
+
+    @property
+    def card(self):
+        """
+        The card property
+        :return: card value
+        """
+        return self._card
+
+    @property
+    def domain(self):
+        """
+        The domain property
+
+        :return: domain value
+        """
+        return self._domain
+
+    @staticmethod
+    def from_string(line):
```

Comment:
May I ask whether those member variables can be wrapped into __init__(), rather than introduce additional static method from_string ?

Reply:
@chunfuwen I'm sorry, I don't understand. Which member variable do you mean could be wrapped into **init()**?

Having `from_string` is a common constructor pattern in python, see for example comment https://github.com/avocado-framework/avocado-vt/pull/2610#discussion_r445400984

If you mean that the `__init__` should have signature `__init__(self, line)` then I prefer not to do that. By having the initialization (with None) and the setting of fields from a string separate, I can still decide to manually define an instance `CryptoDeviceInfoEntry` without the need to first codify the desired entry into a string.